vpdq/cpp/hashing/filehasher.cpp

```diff
@@ -153,271 +394,117 @@ bool hashVideoFile(
-  if (avcodec_open2(codecContext, codec, nullptr) < 0) {
-    std::cerr << "Cannot open codec context" << std::endl;
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
-    return false;
-  }
-
-  // Get the framerate
-  AVRational avframeRate =
-      formatContext->streams[videoStreamIndex]->avg_frame_rate;
-
-  // if avg_frame_rate is 0, fall back to r_frame_rate which is the
-  // lowest framerate with which all timestamps can be represented accurately
-  if (avframeRate.num == 0 || avframeRate.den == 0) {
-    avframeRate = formatContext->streams[videoStreamIndex]->r_frame_rate;
-  }
-
-  double frameRate = static_cast<double>(avframeRate.num) /
-      static_cast<double>(avframeRate.den);
-  if (frameRate == 0) {
-    std::cerr << "Framerate is zero" << std::endl;
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
-    return false;
-  }
-
-  // Create the output frame
-  AVFrame* frame = av_frame_alloc();
-  if (frame == nullptr) {
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
-    return false;
-  }
-
-  // Pixel format for the image passed to PDQ
-  constexpr AVPixelFormat pixelFormat = AV_PIX_FMT_RGB24;
-
-  // Create a frame for resizing and converting the decoded frame to RGB24
-  AVFrame* targetFrame = av_frame_alloc();
-  if (targetFrame == nullptr) {
-    av_frame_free(&frame);
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
-    return false;
-  }
-
-  targetFrame->format = pixelFormat;
-  targetFrame->width = width;
-  targetFrame->height = height;
-
-  if (av_image_alloc(
-          targetFrame->data,
-          targetFrame->linesize,
-          width,
-          height,
-          pixelFormat,
-          1) < 0) {
-    std::cerr << "Cannot allocate target frame" << std::endl;
-    av_frame_free(&targetFrame);
-    av_frame_free(&frame);
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
+  if (downsampleHeight > 0) {
+    video->height = downsampleHeight;
   }
 
   // Create the image rescaler context
-  SwsContext* swsContext = sws_getContext(
-      codecContext->width,
-      codecContext->height,
-      codecContext->pix_fmt,
-      width,
-      height,
-      pixelFormat,
-      SWS_LANCZOS,
+  video->swsContext = sws_getContext(
+      video->codecContext->width,
+      video->codecContext->height,
+      video->codecContext->pix_fmt,
+      video->width,
+      video->height,
+      PIXEL_FORMAT,
+      DOWNSAMPLE_METHOD,
       nullptr,
       nullptr,
       nullptr);
 
-  if (swsContext == nullptr) {
+  if (video->swsContext == nullptr) {
     std::cerr << "Cannot create sws context" << std::endl;
-    av_freep(targetFrame->data);
-    av_frame_free(&targetFrame);
-    av_frame_free(&frame);
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
     return false;
   }
 
   AVPacket* packet = av_packet_alloc();
```

Comment:
Yes, that's a good idea. But packet's are not entirely made simple by the RAII since I'm reusing the same one for reading frames and each iteration its buffer MUST be freed with av_packet_unref(packet); so that might complicate things. I think I might have to use a shared_ptr.